### PR TITLE
HEXDEV-758 - Upgrade Jetty version to 9.4.30.v20200611

### PR DESCRIPTION
**Pull request by Pscheidl**
[HEXDEV-758](https://0xdata.atlassian.net/browse/HEXDEV-758)
**Please note details of the original issue are confidential.**

- Updated Jetty version from series `9.3` to `9.4` - this issue should be fixed for `9.3` series in `9.3.24` as well, yet it's probably a better idea to upgrade to latest `9.4` - as this is the latest and most often maintained.  In `9.4` series, the issue should be fixed in `9.4.11` according to the report - our new version is the newest available at the moment: `9.4.30.v20200611'`.
- Upgrade from Jetty 9.3 to Jetty 9.4 has been done by means of the official [migration guide](https://www.eclipse.org/jetty/documentation/current/upgrading-jetty.html) - There was only a slight change in API calls. Little adjustment in our code was required regarding the `Session Management` section. 
- Include Jetty 9 by default, as Jetty 8 is long deprecated.
- To be absolutely 100 % sure proper version of Jetty 9 is included, I checked the resulting JAR manually.
Screenshot from 2020-07-09 07-45-07

## Before-merge TODO-LIST
- Run all HADOOP tests against this branch (http://mr-0xc1:8080/view/H2O-3/job/h2o-3-hadoop-smoke-pipeline/job/pavel%252Fhexdev-758/)
- Run all Kerberos tests against this branch (http://mr-0xc1:8080/view/H2O-3/job/h2o-3-kerberos-smoke-pipeline/job/pavel%252Fhexdev-758/)
- Run nightlies against this branch (http://mr-0xc1:8080/view/H2O-3/job/h2o-3-nightly-pipeline/job/pavel%252Fhexdev-758/)
- Check with SNYK (it seems to know about the vulnerabilites in Jetty: https://app.snyk.io/vuln/SNYK-JAVA-ORGECLIPSEJETTY-30090)
- Approval of Sparkling Water team is required: @jakubhava @mn-mikke

### CODE REVIEW

File: h2o-jetty-9/build.gradle [hunk 1]
```diff
@@ -1,10 +1,10 @@
 dependencies {
     compile project(":h2o-webserver-iface")
-    compile 'org.eclipse.jetty:jetty-server:9.3.20.v20170531'
-    compile "org.eclipse.jetty:jetty-servlets:9.3.20.v20170531"
-    compile 'org.eclipse.jetty:jetty-jaas:9.3.20.v20170531'
-    compile 'org.eclipse.jetty:jetty-proxy:9.3.20.v20170531'
-    compile 'org.eclipse.jetty:jetty-servlet:9.3.20.v20170531'
+    compile 'org.eclipse.jetty:jetty-server:9.4.30.v20200611'
```
Comment (honzasterba) [inline h2o-jetty-9/build.gradle:3]: I feel like we should put this into a
`ext {
    jetty9version = "4.30"
}`

**Comment by honzasterba** [comment]
Should we also upgrade jetty8 to 8.2.0.v20160908?

**Comment by Pscheidl** [comment]
> Should we also upgrade jetty8 to 8.2.0.v20160908?

Good point. Let's do it in this PR as well. Will do it now. It might break a lot of things, so we might end up doing it separately, as this fix is of high importance for the next release.

**Comment by Pscheidl** [comment]
No problems observed locally. Moved the version to a single place in both 8 and 9 as well. Now let's test it.

File: h2o-jetty-8/build.gradle [hunk 2]
```diff
@@ -2,9 +2,10 @@ description = "Brings Jetty 8 as the HTTP implementation used in H2O server-side
 
 dependencies {
     compile project(":h2o-webserver-iface")
-    compile 'org.eclipse.jetty:jetty-server:8.1.21.v20160908'
-    compile "org.eclipse.jetty:jetty-servlets:8.1.21.v20160908"
-    compile "org.eclipse.jetty:jetty-plus:8.1.21.v20160908"
+    ext { jetty8version = "8.2.0.v20160908" }
```
Comment (honzasterba) [inline h2o-jetty-8/build.gradle:5]: this is usually on the top level of the build.gradle here it is quite hidden

Comment (Pscheidl) [reply h2o-jetty-8/build.gradle:5]: We might put it to properties then to have it at one place :)

File: gradle.properties [hunk 3]
```diff
@@ -61,4 +61,6 @@ defaultHdfsDependency=hadoop-hdfs-client
 # Default Hive version
 defaultHiveExecVersion=1.1.0
 
-defaultWebserverModule=h2o-jetty-8
+defaultWebserverModule=h2o-jetty-9
+jetty8version=8.2.0.v20160908
+jetty9version=9.4.30.v20200611
```
Comment (mn-mikke) [inline gradle.properties:66]: nit: new line

Comment (Pscheidl) [reply gradle.properties]: Thank you, fixed.

**Comment by Pscheidl** [comment]
Re-started tests to run against new master-based branch.

**Comment by Pscheidl** [comment]
Test failures in Hadoop - I guess the issue will be with Jetty defaulting to version 9 even in cases where 8 is required for Hadoop. Same failures repeated for Kerberos tests. Nightlies look good. This should be solved by adjusting our dependencies in the builds.

File: h2o-jetty-9/src/main/java/water/webserver/jetty9/Jetty9WebsocketServlet.java [hunk 4]
```diff
@@ -68,9 +67,30 @@ public void onWebSocketError(Throwable cause) {
...
+
+    /**
+     * Custom in-place socket creator, returning new instance of {@link Jetty9WebsocketHandler}m
+     * which already contains the proper {@link WebsocketServlet} implementation the request is being delegated to.
+     * <p>
+     * This is required, as default {@link WebSocketServletFactory} uses {@link org.eclipse.jetty.util.DecoratedObjectFactory}
+     * to instantiate {@link WebSocketListener} classes. This class is only able to instantiate static classes with 0-arg constructor,
+     * which inner non-static class {@link Jetty9WebsocketHandler} is NOT.
+     */
+    public class H2OWebSocketCreator implements WebSocketCreator {
```
Comment (Pscheidl) [inline h2o-jetty-9/src/main/java/water/webserver/jetty9/Jetty9WebsocketServlet.java:89]: @honzasterba Please review this part. There is a problem with Jetty 9 - it's unable to instantiate `Jetty9WebsocketHandler` which delegates the calls to the proper implementation. It's all described for everyone in the Javadoc in detail - the default factory is only able to instantiate static classes with 0-arg constructors, our class inner non-static. As mentioned in the JavaDoc, each servlet has it's own factory instance - this means the custom `H2OWebSocketCreator` is only assigned to a servlet-local factory, not a global one.

Comment (honzasterba) [reply h2o-jetty-9/src/main/java/water/webserver/jetty9/Jetty9WebsocketServlet.java:89]: this looks like the best solution
knit: H2OWebSocketCreator does not need to be public
self-shaming: I wish I named `water.webserver.iface.WebsocketServlet` differently (with H2O prefix), now it is confusing with jetty's `WebSocketServlet`

Comment (Pscheidl) [reply h2o-jetty-9/src/main/java/water/webserver/jetty9/Jetty9WebsocketServlet.java:89]: Thanks, I'll remove the public keyword once the tests which are in the middle of finishing actually finish. 

I guess this is internal and we can always rename it btw :wink:

Comment (honzasterba) [reply h2o-jetty-9/src/main/java/water/webserver/jetty9/Jetty9WebsocketServlet.java:89]: Good call, will do that after you merge and before we release

**Comment by Pscheidl** [comment]
LDAP is broken on Jetty 9 - looks like it doesn't even make the call to the directory.

**Comment by Pscheidl** [comment]
There are still Kerberos and Hadoop 3.x - based test failures.

**Pull request merged**